src/core/lwt.mli

```diff
@@ -21,502 +22,1785 @@
+    two lines from STDIN concurrently, let's start two waits instead:
+
+{[
+let () =
+  Lwt_main.run begin
+    let three_seconds : unit Lwt.t = Lwt_unix.sleep 3. in
+    let five_seconds : unit Lwt.t = Lwt_unix.sleep 5. in
+    let%lwt () = three_seconds in
+    let%lwt () = Lwt_io.printl "3 seconds passed" in
+    let%lwt () = five_seconds in
+    Lwt_io.printl "Only 2 more seconds passed"
+  end
+
+(* ocamlfind opt -linkpkg -package lwt.ppx,lwt.unix code.ml && ./a.out *)
+]}
+
+    This program takes about five seconds to run. We are still new to [let%lwt],
+    so let's desugar it:
+
+{[
+let () =
+  Lwt_main.run begin
+    let three_seconds : unit Lwt.t = Lwt_unix.sleep 3. in
+    let five_seconds : unit Lwt.t = Lwt_unix.sleep 5. in
+
+    (* Both waits have already been started at this point! *)
+
+    Lwt.bind three_seconds (fun () ->
+      (* This is 3 seconds later. *)
+      Lwt.bind (Lwt_io.printl "3 seconds passed") (fun () ->
+        Lwt.bind five_seconds (fun () ->
+          (* Only 2 seconds were left in the 5-second wait, so
+              this callback runs 2 seconds after the first callback. *)
+          Lwt_io.printl "Only 2 more seconds passed")))
+  end
+
+(* ocamlfind opt -linkpkg -package lwt.unix code.ml && ./a.out *)
+]}
+
+    And that's it! Concurrency in Lwt is simply a matter of whether you start an
+    operation in the callback of another one or not.
+
+    {3 Execution model}
+
+    It's important to understand that promises are a pure-OCaml data type. They
+    don't do any fancy scheduling or I/O. They are just lists of callbacks if
+    pending, and containers for one value if resolved.
+
+    The interesting function is {!Lwt_main.run}. It's a wrapper around
+    {{: http://man7.org/linux/man-pages/man2/select.2.html} [select(2)]},
+    {{: http://man7.org/linux/man-pages/man7/epoll.7.html} [epoll(7)]},
+    {{: https://www.freebsd.org/cgi/man.cgi?query=kqueue&sektion=2}
+    [kqueue(2)]}, or whatever asynchronous API your system provides. Let's take
+    [select(2)] for the example.
+
+    On browsers, the [select(2)] call is inside the JavaScript engine, so you
+    don't call {!Lwt_main.run}. But the logic is still the same!
+
+    Let's use the program from the tutorial that reads two lines as an example.
+    Here it is, again, in its desugared form:
+
+{[
+let () =
+  let p : unit Lwt.t =
+    let line_1_promise : string Lwt.t = Lwt_io.(read_line stdin) in
+    Lwt.bind line_1_promise (fun (line_1 : string) ->
+
+      let line_2_promise : string Lwt.t = Lwt_io.(read_line stdin) in
+      Lwt.bind line_2_promise (fun (line_2 : string) ->
+
+        Lwt_io.printf "%s and %s\n" line_1 line_2))
+  in
+
+  Lwt_main.run p
+
+(* ocamlfind opt -linkpkg -package lwt.unix code.ml && ./a.out *)
+]}
+
+    {!Lwt_main.run} is your program's main I/O loop. You pass it a single
+    top-level promise, and it:
+
+    + Uses [select(2)] to put your process to sleep until the next I/O
+      completes.
```

Comment:
`Lwt_main.run` doesn't execute `p`, because `p` is a promise (a boring data "object," not something that can be executed). If one writes, for example,

```ocaml
Lwt_main.run (Lwt_io.(read_line stdin))
```

it's equivalent, down to the syntactic level, to

```ocaml
let p = Lwt_io.(read_line stdin) in
Lwt_main.run p
```

What happens here is that running `Lwt_io.read_line` sets up the `read`, regardless of the existence of any later call to `Lwt_main.run`, and by the time `Lwt_main.run` is called, there is just a pending promise with nothing at all left to execute (except `select`).

So, indeed, `Lwt_main.run` does first put your process to sleep until the next I/O to complete. This is true in all cases. I specifically want to be clear about this, to avoid the idea that `Lwt_main.run` ever executes anything, *except* after calling `select` and seeing least one I/O has completed (modulo any internal optimizations). I also want to avoid the idea that `p` is special in any way, in the sense of `Lwt_main.run` driving the evaluation of the expression that produces it. That expression is evaluated due to the surrounding semantics of OCaml. The only thing special about `p` that concerns `Lwt_main.run`, is that it acts as a flag for stopping the I/O loop and not calling `select` anymore, and that whether `p` is fulfilled or rejected determines which value `Lwt_main.run` evaluates to, respectively what exception it raises.

Reply:
I guess I should clarify in the docs that `Lwt_main.run` doesn't drive the I/O that is associated with `p` either. That is, it is not because we passed `p` to `Lwt_main.run`, that the I/O "behind" the `read_line` call ran.

Reply:
The explanation in this comment thread is good. Maybe it should be included (with minor edits) into the manual!

Reply:
Yes, thanks for helping to work it out :)

I think a lot of this belongs in the docs of `Lwt_main.run`, but since `Lwt_main` is not up for rewriting yet, maybe I can somehow "park" all of it in Lwt for now. Or just leave a note to revisit the distribution of information between `Lwt` and `Lwt_main` run later.

Reply:
Also, I just realized I gave myself a thumbs-up, lol. I moved it to your comment, where it belongs :)

Reply:
Looking at it more, I decided to save it for addressing later, when more of the docs are written. This is because this page is already trying to be

1. A dry reference for `Lwt`
2. A walkthrough for `Lwt`
3. Examples for `Lwt`
4. A dry reference for `Lwt_main.run`
5. A walkthrough for `Lwt_main.run`
6. Examples for `Lwt_main.run`
7. A few other things
8. Still relatively concise and comprehensible

It's basically not going to satisfy all of these. We will revisit this once there are more pages to tease this apart into.